contracts/SwapUtilsV2.sol

```diff
@@ -0,0 +1,1048 @@
+     * the same cardinality as the pool.
+     * @param d the stableswap invariant
+     * @return the price of the token, in the same precision as in xp
+     */
+    function getYD(
+        uint256 a,
+        uint8 tokenIndex,
+        uint256[] memory xp,
+        uint256 d
+    ) internal pure returns (uint256) {
+        uint256 numTokens = xp.length;
+        require(tokenIndex < numTokens, "Token not found");
+
+        uint256 c = d;
+        uint256 s;
+        uint256 nA = a * (numTokens);
+
+        for (uint256 i = 0; i < numTokens; i++) {
+            if (i != tokenIndex) {
+                s = s + (xp[i]);
+                c = (c * (d)) / (xp[i] * (numTokens));
+                // If we were to protect the division loss we would have to keep the denominator separate
+                // and divide at the end. However this leads to overflow with large numTokens or/and D.
+                // c = c * D * D * D * ... overflow!
+            }
+        }
+        c = (c * (d) * (AmplificationUtilsV2.A_PRECISION)) / (nA * (numTokens));
+
+        uint256 b = s + ((d * (AmplificationUtilsV2.A_PRECISION)) / (nA));
+        uint256 yPrev;
+        uint256 y = d;
+        for (uint256 i = 0; i < MAX_LOOP_LIMIT; i++) {
+            yPrev = y;
+            y = y * (y) + (c) / (y * (2) + (b) - (d));
+            if (y.within1(yPrev)) {
+                return y;
+            }
+        }
+        revert("Approximation did not converge");
+    }
+
+    /**
+     * @notice Get D, the StableSwap invariant, based on a set of balances and a particular A.
+     * @param xp a precision-adjusted set of pool balances. Array should be the same cardinality
+     * as the pool.
+     * @param a the amplification coefficient * n * (n - 1) in A_PRECISION.
+     * See the StableSwap paper for details
+     * @return the invariant, at the precision of the pool
+     */
+    function getD(uint256[] memory xp, uint256 a)
+        internal
+        pure
+        returns (uint256)
+    {
+        uint256 numTokens = xp.length;
+        uint256 s;
+        for (uint256 i = 0; i < numTokens; i++) {
+            s = s + (xp[i]);
+        }
+        if (s == 0) {
+            return 0;
+        }
+
+        uint256 prevD;
+        uint256 d = s;
+        uint256 nA = a * (numTokens);
+
+        for (uint256 i = 0; i < MAX_LOOP_LIMIT; i++) {
+            uint256 dP = d;
+            for (uint256 j = 0; j < numTokens; j++) {
+                dP = (dP * (d)) / (xp[j] * (numTokens));
+                // If we were to protect the division loss we would have to keep the denominator separate
+                // and divide at the end. However this leads to overflow with large numTokens or/and D.
+                // dP = dP * D * D * D * ... overflow!
+            }
+            prevD = d;
+            d =
+                (nA * (s)) /
+                (AmplificationUtilsV2.A_PRECISION) +
+                ((dP * (numTokens)) * (d)) /
```

Comment:
Doesn't that `*(d)` belong one level further out on the paranthesis? Meaning, the whole expression gets multiplied by `d` , not just the `(dP * (numTokens))`